add the start on new calculation method for getisord statistics

ljwolf opened the pull request:
This starts addressing #121 

# Goals
1. keep the computation of the *statistic* constant by configuring/modifying the weights to reflect the correct combination of `star` and `transform`. 
2. make this amenable to use in the new `crand.py` module
3. keep argument compatibility with old `G_Local` class
4. provide the same results & p-values as old `G_Local` class. 

## Optional
- move to a new-form statistic & add deprecation warnings to the deprecating properties? 

## Realizations
*(Just to keep track of stuff I realized during writing this PR)*
- Methods that mutate the source object (like `G_Local.calc`) can make it difficult to assess whether (a) the same attributes are built onto the object in the same manner for all branches of code and (b) the precise sequence of where `self` attributes are created. Focusing on functions with consistent return values is easier, and using [dataclasses](https://docs.python.org/3/library/dataclasses.html) might be easier still. A solution might be to implement stats as dataclasses with a `.fit()` method that takes init input as the default, but can override that with anything passed to fit. 

# Remaining to do
- [x] remove the older calculation code
- [x] wire up `__calc2` to the result containers in the class
- [x] move the `_infer_...` call into the init. 
- [x] write `_infer_...` tests & docstrings
- [x] hook up to `crand.py`, which includes
  - [x] writing and validating the `_local_gi` functions
  - [x] introducing the `keep_simulations` argument

ljwolf commented on esda/getisord.py:
```diff
@@ -518,5 +557,88 @@ def by_col(
+        elif transform.lower() == "r":
+            # This warning is probably better as a documentation instead of a warning.
+            # warnings.warn(
+            #     "Gi* requested, but no weights are on the diagonal and"
```
This also needs to get added to the documentation.

ljwolf commented on esda/getisord.py:
```diff
@@ -518,5 +557,88 @@ def by_col(
+        numpy.testing.assert_allclose(empirical_variance, test.s2)
+
+    # Also check that the None configuration works
+    test = esda.getisord.G_Local(df.GI89, w, star=None)
```
This will fail right now, but the intent is to eventually test the `star==None`, which should be equivalent to `star==False` for this configuration. 
Then, fill the weights diagonal with 1 and check that None returns results akin to star==True.

ljwolf commented:
Now I run into an interesting bug: when `n_jobs > n_obs`, the [`hstack` to collect results from the workers fails](https://github.com/pysal/esda/blob/master/esda/crand.py#L466) because some are sent no jobs. I am not sure how to deal with this yet, other than a hard check to set `n_jobs  <= n_obs`.

ljwolf commented:
I have resolved the small dataset issue. But, now I am realizing a somewhat larger issue. 

In the conditional randomization, we're using the "flat" weights that come from `w.sparse` to compute local statistics in a fast way. 
But, the weights in that set are stored in row-order. This means that *if there is a self-neighbor*, *that weight can occur anywhere in the weight set for that observation. 

This poses an issue for Gi, since we need to randomize the surrounding site values, but keep the site value fixed. 

This is also a problem in Moran statistics when the user inputs a w with self-weight (we're not checking that the w has no self-neighbor and, if it does, we don't hold the self-site constant).

darribas commented:
Mmm... I see the technical point but, conceptually, people should _not_ be passing weights with self neighbors (except on kernel weights methods like HAC, etc. My sense is this is something we should definitely document (like we do with the assumption that `W` and `y` are aligned, and maybe we could build a general checker for sparse weights that performs this kind of check (and potentially others like islands, etc.) and hook it into this class (and potentially any other that could benefit). If the check is expensive, we could add a preset parameter `check_w=True` to turn it off?

ljwolf commented:
> conceptually, people should *not* be passing weights with self-neighbors

I agree for Moran, since covariation between *z* and *lag(z)* will undoubtedly be stronger if *zi* is used in *lag(z)_i*... to pass self-weighted **W** blurs the point of the Moran statistic. That said, [Anselin (1995, p. 98)](https://doi.org/10.1111/j.1538-4632.1995.tb00338.x) only says the diagonal is zero "by convention" (suggesting it's not a necessity) and [Ord & Getis (1995, p289)](https://doi.org/10.1111/j.1538-4632.1995.tb00912.x) explicitly relax this and derive an expression for *I* in terms of *Gi*-star, where *wii ≠ 0*.

To support fast inference for *Gi*-star-type statistics (I believe this also includes the LOSH statistic @jeffcsauer is working on), we need to be able to process self-weights in the crand engine. The alternative for *Gi*-star would be to force *w.j* to be the same within each *i*, making any distance-decay based weight impossible. This is generally not a condition of the *Gi*-family statistics, and (given Ord's construction) I think is unnecessarily restrictive for the Moran family, too. 

Since I have already implemented the extension to allow for self-weights locally (it ended up being quite simple), I suggest we 
go for it & ensure the timings do not degrade? And, if not, should we warn if `Moran_Local` sees a weights matrix with self-neighbors? 

Performance concerns might include:
1. zeroing the diagonal of the weights matrix (may be resolved by converting to `lil` format)
2. allocating a weights vector for each observation, in each chunk (may be resolved by allocating one vector of size `max_cardinality` in each chunk).

ljwolf commented:
xref #86

ljwolf commented:
Single threaded performance is comparable. I'm getting around 1 second for NAT, which is the same as before handling self-neighbors.

ljwolf commented:
OK, I believe this is as close as it can get. Running tests locally, I get  *p-values* that are within 1 simulation of significance using the same seed. In theory, these should be identical, but I this passes our semantic tests we've used before.

ljwolf commented:
Now there are conflicts in `crand.py` that need to be addressed before merging. Should be easy to resolve.

ljwolf commented on esda/getisord.py:
```diff
@@ -382,19 +385,32 @@ def __init__(
         self.permutations = permutations
         self.star = star
         self.calc()
         self.p_norm = np.array([1 - stats.norm.cdf(np.abs(i)) for i in self.Zs])
```
This needs to follow line 420, as per #181

ljwolf commented:
OK, I've rebased this again and addressed emergent issues with the new estimator classes (local Geary, join counts, etc). 

Tests pass, documentation is improved for the star option... 

:shipit:

ljwolf commented:
I'd like to get this merged, and then cut a release so that it makes it into the July PySAL.

jGaboardi reviewed (approved):
Not sure if changes are necessary for the comments I made, I'll leave that to the "committee." Otherwise, looks good to me.

ljwolf commented on esda/getisord.py:
```diff
@@ -530,22 +552,115 @@ def by_col(
+    star = (not zero_diagonal) if star is None else star
+
+    # Want zero diagonal but do not have it
+    if (not zero_diagonal) & (star is False):
```
It would, but this is where the the trickiness of this API comes in. There are *three* correct values for `star`: `True`/`False`/`float`. If `star` is a float, we need to use it as the default value for the self-weight, but if it's `True`, we need to set a default.

jGaboardi replied:
Ah. Def something that I missed.

ljwolf commented:
I suppose the perspective that'd be useful here is to know if the `_infer` function makes sense... 

If `star` is True, we need to set the weights intelligently: for binary weights, we can just set the self-weight to 1, but if the weights are row-standardised, it's hard to know what to set as the default value... If you pass `star=2, transform='r'`, for example, you're saying that you want to set the self-weight to 2, then row-standardise, so that the self-join should be worth two other joins.

knaaptime reviewed (approved):
wish i had more to offer, but been through the code several times now and looks great
